File: pe-lwjgl-binding/src/main/java/org/saar/lwjgl/opengl/shaders/uniforms/Vec3UniformValue.java
package org.saar.lwjgl.opengl.shaders.uniforms;

import org.joml.Vector3f;
import org.joml.Vector3fc;
import org.saar.maths.utils.Vector3;

public class Vec3UniformValue extends Vec3Uniform implements UniformValue<Vector3fc> {

    private final String name;

    private final Vector3f value = Vector3.create();

    public Vec3UniformValue(String name) {
        this.name = name;
    }

    @Override
    public final Vector3fc getUniformValue() {
        return getValue();
    }

    @Override
    public final String getName() {
        return this.name;
    }

    @Override
    public final Vector3f getValue() {
        return this.value;
    }

    @Override
    public final void setValue(Vector3fc value) {
        getValue().set(value);
    }
}

File: pe-lwjgl-binding/src/main/java/org/saar/lwjgl/opengl/shaders/uniforms/Vec3UniformValue.kt
package org.saar.lwjgl.opengl.shaders.uniforms

import org.joml.Vector3fc
import org.saar.maths.JomlDelegates

class Vec3UniformValue(private val name: String) : Vec3Uniform(), UniformValue<Vector3fc> {

    private var vector: Vector3fc by JomlDelegates.CachedVector3f()

    override fun getUniformValue() = this.value

    override fun getName() = this.name

    override fun getValue() = this.vector

    override fun setValue(value: Vector3fc) {
        this.vector = value
    }
}